### ISSUE TITLE
MPI_Allgather_init failures wtih MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=sched_recursive_doubling, which is the default

### ISSUE THREAD

**Issue by jeffhammond** [pmodels/mpich#6423]
I see failures in the MPICH test `p_allgather` due to `MPI_COMM_WORLD` being an invalid communicator, which happens only for 2 or 4 processes, from the same rank (0 or 3).  With 8 and 16 processes, it deadlocks.  For non-powers-of-two, I don't see problems, as expected from what follows.

I find that
```
MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=sched_ring
MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=sched_brucks
MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=tsp_ring
MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=tsp_brucks
MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=tsp_recexch_doubling
MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=tsp_recexch_halving
```
resolve the problem, while
```
MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=sched_recursive_doubling
MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=auto
```
show the error.

I read down to `MPIR_Iallgather_intra_sched_recursive_doubling` but I cannot see any obvious problems, although it is clear that this function is only used for powers-of-two, so it makes sense that it correlates with the failures I observe.

# Details

```
jhammond@nuclear:/tmp$ /opt/mpich/gcc/debug-ch4ucx/bin/mpichversion
MPICH Version:      4.2a1
MPICH Release date: unreleased development copy
MPICH ABI:          0:0:0
MPICH Device:       ch4:ucx
MPICH configure:    CC=gcc CXX=g++ FC=gfortran --enable-g=all,handle,handlealloc,mem,meminit,memarena --enable-versioning --enable-debuginfo --without-ze --prefix=/opt/mpich/gcc/debug-ch4ucx --with-device=ch4:ucx
MPICH CC:           gcc    -g -O2
MPICH CXX:          g++   -g -O2
MPICH F77:          gfortran   -g -O2
MPICH FC:           gfortran   -g -O2
```

```c
/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
/*
 *  (C) 2010 by Argonne National Laboratory.
 *      See COPYRIGHT in top-level directory.
 */

/* This is a very weak sanity test that persistenr broadcast collectives specified by
 * proposed MPI-4 is present in the library and take arguments as expected.  This test
 * does not check for progress or matching issues. */

#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

#define NUM_INTS (2)

#define my_assert(cond_)                                                  \
    do {                                                                  \
        if (!(cond_)) {                                                   \
            fprintf(stderr, "assertion (%s) failed, aborting\n", #cond_); \
            MPI_Abort(MPI_COMM_WORLD, 1);                                 \
        }                                                                 \
    } while (0)

int main(int argc, char **argv)
{
    int errs = 0;
    int i;
    int rank, size;
    int *sendbuf = NULL;
    int *recvbuf = NULL;
    MPI_Comm comm;
    int count = 10;
    int j = 0;

    MPI_Init(&argc, &argv);

    comm = MPI_COMM_WORLD;

    MPI_Comm_size(comm, &size);
    MPI_Comm_rank(comm, &rank);

    sendbuf = malloc(NUM_INTS * sizeof(int));
    recvbuf = malloc(size * NUM_INTS * sizeof(int));
    my_assert(sendbuf);
    my_assert(recvbuf);

    for (i = 0; i < NUM_INTS; i++) {
        sendbuf[i] = rank * NUM_INTS + i;
    }

    for (i = 0; i < (size * NUM_INTS); i++) {
        recvbuf[i] = -1;
    }

    MPI_Info info;
    MPI_Info_create(&info);
    MPI_Request req;

    MPI_Allgather_init(sendbuf, NUM_INTS, MPI_INT, recvbuf, NUM_INTS, MPI_INT, comm, info, &req);

    for (i = 0; i < count; i++) {
        MPI_Start(&req);
        MPI_Wait(&req, MPI_STATUS_IGNORE);
        /* Check the results */
        for (j = 0; j < (size * NUM_INTS); j++) {
            if (recvbuf[j] != j) {
                errs++;
                if (errs < 10) {
                    fprintf(stderr, "rbuf[%d] = %d expected %d\n", j, recvbuf[j], j);
                }
            }
        }
    }

    MPI_Request_free(&req);

    MPI_Info_free(&info);

    free(sendbuf);
    free(recvbuf);

    MPI_Barrier(MPI_COMM_WORLD);

    MPI_Finalize();
    return errs;
}
```

```
jhammond@nuclear:/tmp$ rm raw_p_allgather.x ; /opt/mpich/gcc/debug-ch4ucx/bin/mpicc -g -Os raw_p_allgather.c -o raw_p_allgather.x && MPIR_CVAR_IALLGATHER_DEVICE_COLLECTIVE=0 MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=sched_recursive_doubling /opt/mpich/gcc/debug-ch4ucx/bin/mpirun -l -n 2  ./raw_p_allgather.x || echo FAIL
Authorization required, but no authorization protocol specified
[0] Abort(739945231) on node 0 (rank 0 in comm 0): Fatal error in internal_Wait: Other MPI error, error stack:
[0] internal_Wait(103)..............: MPI_Wait(request=0x7ffec61ecb74, status=0x1) failed
[0] internal_Wait(92)...............:
[0] MPIDU_Sched_progress_state(1183): Invalid communicator
FAIL
jhammond@nuclear:/tmp$ rm raw_p_allgather.x ; /opt/mpich/gcc/debug-ch4ucx/bin/mpicc -g -Os raw_p_allgather.c -o raw_p_allgather.x && MPIR_CVAR_IALLGATHER_DEVICE_COLLECTIVE=0 MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=sched_recursive_doubling /opt/mpich/gcc/debug-ch4ucx/bin/mpirun -l -n 4  ./raw_p_allgather.x || echo FAIL
Authorization required, but no authorization protocol specified
[3] Abort(807054095) on node 3 (rank 3 in comm 0): Fatal error in internal_Wait: Other MPI error, error stack:
[3] internal_Wait(103)..............: MPI_Wait(request=0x7ffeb277c5b4, status=0x1) failed
[3] internal_Wait(92)...............:
[3] MPIDU_Sched_progress_state(1183): Invalid communicator
FAIL
```

**Comment by jeffhammond** [comment]
I don't know if it means anything, but I ran with `MPIR_CVAR_COLL_SCHED_DUMP=1` and see `MPI_UNDEFINED` a bunch, which I ignorantly speculate _might_ be an indication of a problem, only because the `sched_brucks` contains none of them.

```
[1] &s->entries[1]=0x55e9d9cef430
[1]     s->entries[1].type=SEND
[1]     s->entries[1].status=1
[1]     s->entries[1].is_barrier=FALSE
[1]             Send: -32766 of type 4c000405 from 0
[1]              from buff: 0x55e9d9d2a648
```

**Comment by jeffhammond** [comment]
I last compiled from scratch on February 20.  I did a partial update and didn't see any change, but I am doing a completely clean build now to verify.

**Comment by hzhou** [comment]
I can't reproduce the failure.
The error message is from `MPI_ERR_OTHER`. A previous sloppy coder used a misleading error message. It is nothing to do with "Invalid communicator", but there is little clue on what was the error either.

The schedule was sending `MPI_UNDEFINED` count of `MPI_INT`, that is definitely wrong. I'll wait for your clean build verification.

**Comment by jeffhammond** [comment]
2f3a6ec5963fce27fb34ff79147d7fffa707c1ac still shows it for me with 

```
git clean -dfx ; git pull ; ./autogen.sh -yaksa-depth=0 && mkdir -p build && cd build && ../configure CC=gcc CXX=g++ FC=gfortran --enable-g=all,handle,handlealloc,mem,meminit,memarena --enable-versioning --enable-debuginfo --without-ze --prefix=/opt/mpich/gcc/debug-ch4ucx --with-device=ch4:ucx && make -j8 install
```

**Comment by jeffhammond** [comment]
I am going to try on a completely different system to determine if the machine that shows this bug is suffering from demonic possession.

**Comment by jeffhammond** [comment]
It's reproducible on my end.  I ran on an AArch64 system with Ubuntu 20, whereas the first one was x86_64 with Ubuntu 22.  More importantly, this machine was off throughout the development of Mukautuva so it's impossible for anything I might have done there to have impacted this.

I wonder if the full debug set I use is part of it.  Are you trying to reproduce with the exact same configure command?

```
jhammond@orin:/samsung/MPICH/git/build$ git checkout main ; git reset --hard ; git checkout . ;  git clean -dfx ; git pull ; ./autogen.sh -yaksa-depth=0 && mkdir -p build && cd build && ../configure CC=gcc CXX=g++ FC=gfortran --enable-g=all,handle,handlealloc,mem,meminit,memarena --enable-versioning --enable-debuginfo --without-ze --prefix=/samsung/MPICH/dev-gcc-debug-ch4ucx --with-device=ch4:ucx && make -j8 install
```

```
jhammond@orin:/samsung/MPICH/git/build$ /samsung/MPICH/dev-gcc-debug-ch4ucx/bin/mpicc raw_pallgather.c && MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=sched_brucks /samsung/MPICH/dev-gcc-debug-ch4ucx/bin/mpirun -n 4 ./a.out && echo OK
OK
jhammond@orin:/samsung/MPICH/git/build$ /samsung/MPICH/dev-gcc-debug-ch4ucx/bin/mpicc raw_pallgather.c && MPIR_CVAR_IALLGATHER_INTRA_ALGORITHM=sched_recursive_doubling /samsung/MPICH/dev-gcc-debug-ch4ucx/bin/mpirun -n 4 ./a.out
Abort(1008380687) on node 3 (rank 3 in comm 0): Fatal error in internal_Wait: Other MPI error, error stack:
internal_Wait(103)..............: MPI_Wait(request=0xffffc7800a00, status=0x1) failed
internal_Wait(92)...............:
MPIDU_Sched_progress_state(1183): Invalid communicator
```

**Comment by hzhou** [comment]
Yes, I reproduced it with the full debug settings. Now let me investigate...

**Comment by hzhou** [comment]
I found the bug. When we implemented persistent collectives, we did not study individual algorithms. This particular algorithm uses a state to track counts between algorithm step and we are missing a state reset between iterations of MPI_Start. The bug didn't show up normally because sending more data don't always cause the issue as long as the garbage is finite and are ignored. In particular, when receiving more data than prepared, it raise TRUNCATION error and leave the status object unset. And when turning on the debug flags, the unset fields start contain non-zero garbages and causing the issue we observe here.

TODOs:
1. reset states between MPI_Start
2. Add CVAR to facilitate debugging non-blocking collectives
3. We should not allow loading status count fields in the case of truncation error
